core/src/SYCL/Kokkos_SYCL_Space.cpp

```diff
@@ -0,0 +1,409 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_SYCL_Space.hpp>
+#include <Kokkos_HostSpace.hpp>
+#include <impl/Kokkos_Profiling.hpp>
+#include <SYCL/Kokkos_SYCL_Instance.hpp>
+#include <impl/Kokkos_MemorySpace.hpp>
+
+/*--------------------------------------------------------------------------*/
+/*--------------------------------------------------------------------------*/
+namespace Kokkos {
+namespace Impl {
+namespace {
+void USM_memcpy(cl::sycl::queue& q, void* dst, const void* src, size_t n) {
+  cl::sycl::event e = q.memcpy(dst, src, n);
+  e.wait();
+}
+
+void USM_memcpy(Kokkos::Experimental::Impl::SYCLInternal& space, void* dst,
+                const void* src, size_t n) {
+  USM_memcpy(*space.m_queue, dst, src, n);
+}
+
+void USM_memcpy(void* dst, const void* src, size_t n) {
+  USM_memcpy(Kokkos::Experimental::Impl::SYCLInternal::singleton(), dst, src,
+             n);
+}
+}  // namespace
+
+DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
+         Kokkos::Experimental::SYCLDeviceUSMSpace, Kokkos::Experimental::SYCL>::
+    DeepCopy(const Kokkos::Experimental::SYCL& instance, void* dst,
+             const void* src, size_t n) {
+  USM_memcpy(*instance.impl_internal_space_instance(), dst, src, n);
+}
+
+DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace,
+         Kokkos::Experimental::SYCLDeviceUSMSpace,
+         Kokkos::Experimental::SYCL>::DeepCopy(void* dst, const void* src,
+                                               size_t n) {
+  USM_memcpy(dst, src, n);
+}
+
+DeepCopy<HostSpace, Kokkos::Experimental::SYCLDeviceUSMSpace,
+         Kokkos::Experimental::SYCL>::DeepCopy(const Kokkos::Experimental::SYCL&
+                                                   instance,
+                                               void* dst, const void* src,
+                                               size_t n) {
+  USM_memcpy(*instance.impl_internal_space_instance(), dst, src, n);
+}
+
+DeepCopy<Kokkos::HostSpace, Kokkos::Experimental::SYCLDeviceUSMSpace,
+         Kokkos::Experimental::SYCL>::DeepCopy(void* dst, const void* src,
+                                               size_t n) {
+  USM_memcpy(dst, src, n);
```

Comment:
Also supposed to be synchronous.

Reply:
It seems to be synchronous (see Line 58)